pkg/middleware/middleware.go

```diff
@@ -0,0 +1,124 @@
+	// UsernameKey is the context key for preferred_username claim
+	UsernameKey = "username"
+	// EmailKey is the context key for email claim
+	EmailKey = "email"
+	// JWTClaimsKey is the context key for the claims struct
+	JWTClaimsKey = "jwtClaims"
+)
+
+// JWTMiddleware is the JWT token validation middleware
+type JWTMiddleware struct {
+	config      *configuration.Registry
+	logger      *log.Logger
+	keyManager  *auth.KeyManager
+	tokenParser *auth.TokenParser
+}
+
+// NewAuthMiddleware returns a new middleware for JWT authentication
+func NewAuthMiddleware(logger *log.Logger, config *configuration.Registry) (*JWTMiddleware, error) {
+	if logger == nil || config == nil {
+		return nil, errors.New("missing parameters for NewAuthMiddleware")
+	}
+	// wire up the key and token management
+	keyManagerInstance, err := auth.DefaultKeyManager()
+	if err != nil {
+		return nil, err
+	}
+	tokenParserInstance, err := auth.DefaultTokenParser()
+	if err != nil {
+		return nil, err
+	}
+	return &JWTMiddleware{
+		logger:      logger,
+		config:      config,
+		keyManager:  keyManagerInstance,
+		tokenParser: tokenParserInstance,
+	}, nil
+}
+
+func (m *JWTMiddleware) extractToken(c *gin.Context) (string, error) {
+	// token lookup order: header: Authorization, query: token
+	// try header field "Authorization" (will be "" when n/a)
+	headerToken := c.GetHeader("Authorization")
+	if headerToken != "" {
+		if strings.HasPrefix(headerToken, "Bearer ") {
+			// it is a bearer token, split it up and return it
+			s := strings.Fields(headerToken)
+			if len(s) == 2 {
+				return s[1], nil
+			}
+			// we're failing fast here, if there is an Authorization header, it is used or it fails
+			return "", errors.New("found bearer token header, but no token:" + headerToken)
+		}
+		// see above, failing fast
+		return "", errors.New("found unknown authorization header:" + headerToken)
+	}
+	// next, try GET param "token" (will return "" if n/a)
+	paramToken := c.Query("token")
+	if paramToken != "" {
+		return paramToken, nil
+	}
+	return "", errors.New("no token found")
+}
+
+func (m *JWTMiddleware) respondWithError(c *gin.Context, code int, message interface{}) {
+	c.AbortWithStatusJSON(code, gin.H{"error": message})
+}
+
+// HandlerFunc returns the HanderFunc.
+func (m *JWTMiddleware) HandlerFunc() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		// check if we have a token
+		tokenStr, err := m.extractToken(c)
+		if err != nil {
+			m.respondWithError(c, http.StatusForbidden, err.Error())
+			c.Abort()
+			return
+		}
+		// next, check the token
+		token, err := m.tokenParser.FromString(tokenStr)
+		if err != nil {
+			m.respondWithError(c, http.StatusForbidden, err.Error())
+			c.Abort()
+			return
+		}
+		// validate time claims
+		if token.Valid() != nil {
+			m.respondWithError(c, http.StatusForbidden, "token has invalid time claims")
+			c.Abort()
+			return
+		}
+		// check if we have the needed claims for username and email
+		if token.Username == "" {
```

Comment:
We do need to check Username, email (and also sub which is missing here) but it's already checked by the token parser - https://github.com/codeready-toolchain/registration-service/blob/4edbec2c9dbed84a985618392f2f7c5a45a36fa4/pkg/auth/tokenparser.go#L63-L75
So, all these claim checks are redundant here, aren't they?

Reply:
I usually like to have all components check their "owning" scope independently so there is no dependency on checks of vital states outside of the component. But yeah, removed in ceca752, added comment.

Reply:
Then add a test for that vital state ;) So if the dependency behavior unexpectedly changes the test will catch it.
In other words it makes sense to have some redundancy in this tests. But not in implementation.

Reply:
But this is runtime, not test scope. Anyways, it is checked in tokenparser only now.